pkg/runner/test_runner.go

```diff
@@ -6,282 +6,796 @@
+const resumeStateStructVersion = 1
+
+// Run is the main enty point of the runner.
+func (tr *TestRunner) Run(
+	ctx statectx.Context,
+	t *test.Test, targets []*target.Target,
+	jobID types.JobID, runID types.RunID,
+	resumeState []byte) ([]byte, error) {
+
+	// Set up logger
+	rootLog := logging.GetLogger("pkg/runner")
+	fields := make(map[string]interface{})
+	fields["jobid"] = jobID
+	fields["runid"] = runID
+	rootLog = logging.AddFields(rootLog, fields)
+	tr.log = logging.AddField(rootLog, "phase", "run")
+
+	tr.log.Debugf("== test runner starting job %d, run %d", jobID, runID)
+	resumeState, err := tr.run(ctx, t, targets, jobID, runID, resumeState)
+	tr.log.Debugf("== test runner finished job %d, run %d, err: %v", jobID, runID, err)
+	return resumeState, err
 }
 
-func (w *targetWriter) writeTimeout(ctx context.Context, ch chan<- *target.Target, target *target.Target, timeout time.Duration) error {
-	w.log.Debugf("writing target %+v, timeout %v", target, timeout)
-	start := time.Now()
+func (tr *TestRunner) run(
+	ctx statectx.Context,
+	t *test.Test, targets []*target.Target,
+	jobID types.JobID, runID types.RunID,
+	resumeState []byte) ([]byte, error) {
+
+	// Peel off contexts used for steps and target handlers.
+	stepCtx, _, stepCancel := statectx.WithParent(ctx)
+	defer stepCancel()
+	targetCtx, _, targetCancel := statectx.WithParent(ctx)
+	defer targetCancel()
+
+	// Set up the pipeline
+	for i, sb := range t.TestStepsBundles {
+		tr.steps = append(tr.steps, &stepState{
+			stepIndex: i,
+			sb:        sb,
+			inCh:      make(chan *target.Target),
+			outCh:     make(chan *target.Target),
+			errCh:     make(chan cerrors.TargetError),
+			ev: storage.NewTestEventEmitter(testevent.Header{
+				JobID:         jobID,
+				RunID:         runID,
+				TestName:      t.Name,
+				TestStepLabel: sb.TestStepLabel,
+			}),
+			tgtDone: make(map[*target.Target]bool),
+			log:     logging.AddField(tr.log, "step", sb.TestStepLabel),
+		})
+		// Step handlers will be started from target handlers as targets reach them.
+	}
+
+	// Set up the targets
+	tr.targets = make(map[string]*targetState)
+	// If we have target state to resume, do it now.
+	if len(resumeState) > 0 {
+		tr.log.Debugf("Attempting to resume from state: %s", string(resumeState))
+		var rs resumeStateStruct
+		if err := json.Unmarshal(resumeState, &rs); err != nil {
+			return nil, fmt.Errorf("invalid resume state: %w", err)
+		}
+		if rs.Version != resumeStateStructVersion {
+			return nil, fmt.Errorf("incompatible resume state version %d (want %d)",
+				rs.Version, resumeStateStructVersion)
+		}
+		if rs.JobID != jobID {
+			return nil, fmt.Errorf("wrong resume state, job id %d (want %d)", rs.JobID, jobID)
+		}
+		tr.targets = rs.Targets
+	}
+	// Initialize remaining fields of the target structures,
+	// build the map and kick off target processing.
+	for _, tgt := range targets {
+		ts := tr.targets[tgt.ID]
+		if ts == nil {
+			ts = &targetState{
```

Comment:
This new `targetState` doesn't get backported in `tr.targets`?

EDIT: nevermind, it happens down below.